### PR TITLE
bootstrap: Make `./x test compiler` actually run the compiler unit tests

### LINKED ISSUE #134916: Running `./x test compiler` only runs unit tests for `rustc_codegen_cranelift`
In theory, this command should run `cargo test` for all of the compiler crates in `compiler/`:

```bash
# (stage 0 isn't essential, but it's faster than using a later stage)
./x test compiler --stage=0
```

Instead, it just prints this message and exits successfully:

```text
cranelift not in rust.codegen-backends. skipping
```

---

This seems to be because `test::CodegenCranelift::should_run` registers `compiler/rustc_codegen_cranelift` as its path, which has the side-effect of hijacking `compiler` in bootstrap's wacky path-resolution logic. That prevents `test::CrateLibrustc` from running, which is what would run the unit tests for the vast majority of compiler crates.

The same problem is also theoretically present for `test::CodegenGCC`, though currently the cranelift step takes priority for whatever reason.

### PR DESCRIPTION
**Pull request by Zalathar**
Fixes #134916.

### CODE REVIEW

File: src/bootstrap/src/core/builder/tests.rs [hunk 1]
```diff
@@ -786,3 +786,21 @@ mod sysroot_target_dirs {
...
+    let cmd = &["test", "compiler"].map(str::to_owned);
+    let config = configure_with_args(cmd, &[TEST_TRIPLE_1], &[TEST_TRIPLE_1]);
+    let cache = run_build(&config.paths.clone(), config);
+
+    let compiler = cache.contains::<test::CrateLibrustc>();
+    let cranelift = cache.contains::<test::CodegenCranelift>();
+    let gcc = cache.contains::<test::CodegenGCC>();
+
+    assert_eq!((compiler, cranelift, gcc), (true, false, false));
+}
```
Comment (Zalathar) [inline src/bootstrap/src/core/builder/tests.rs:806]: I have manually verified that this unit test fails before the change, and succeeds after the change.

**Comment by Zalathar** [comment]
Explanation of the fix:

When bootstrap tries to figure out which step is responsible for each command-line argument, it only allows one step to claim any argument. So to force `test::CrateLibrustc` to take priority over the codegen backend steps, we just need to make sure it appears before them in the list of test steps.

This is the code that allows steps to claim arguments:

https://github.com/rust-lang/rust/blob/c1566141b6ed24eaa11075aa7dfe6511fdf32231/src/bootstrap/src/core/builder/mod.rs#L489-L509

File: src/bootstrap/src/core/builder/mod.rs [hunk 2]
```diff
@@ -922,13 +922,13 @@ impl<'a> Builder<'a> {
...
                 test::UiFullDeps,
-                test::CodegenCranelift,
-                test::CodegenGCC,
                 test::Rustdoc,
                 test::CoverageRunRustdoc,
                 test::Pretty,
                 test::Crate,
                 test::CrateLibrustc,
+                test::CodegenCranelift,
+                test::CodegenGCC,
```
Comment (onur-ozkan) [inline src/bootstrap/src/core/builder/mod.rs:931]: Adding an explanatory comment would be great similar to the one at https://github.com/rust-lang/rust/blob/be32f13a44e5bc588ecb5289bfd6305d561fa26e/src/bootstrap/src/core/builder/mod.rs#L1029-L1035

File: src/bootstrap/src/core/builder/tests.rs (hunk 1 above)
Comment (Zalathar) [reply src/bootstrap/src/core/builder/tests.rs:806]: I also tried to write a unit test for `./x test rustc_codegen_cranelift` and `./x test compiler/rustc_codegen_cranelift`, to make sure this PR doesn't break them. But so far I haven't managed to get those tests to work, so for now I've just verified them by hand.

**Comment by jieyouxu** [comment]
FTR, in my investigation, the root problem is potentially the concept of the `PathSet` itself. It admits several issues (some of these are over-simplifications and may be inaccurate or imprecise, but is sufficient to demonstrate the complexity and the issue at hand):

- We conflate *aliases* with on-disk paths -- it's especially weird if there is an on-disk directory that shares the same name as the *alias*. (Example: `compiler`).
- Test suites are handled differently versus other test-ish/non-test `Step`s, where test suites have "run-all" semantics but non-test-suite `Step`s have "run-once-consume-a-path" semantics.
- There's some very weird interactions with `--skip`/`--exclude` when it comes to test suites versus *individual test files* under each test suite. This is not bootstrap's fault (alone), though -- it has to do with how libtest (through `cargo test` or wrapped through `compiletest`) does test filtering based on a substring match on test names that are usually constructed based on the test path *verbatim* (unless you path `--exact`, in which case I think a path prefix doesn't work).
- Before https://github.com/rust-lang/rust/pull/133492, `PathSet::check` matches on *suffixes* (FTR, while this saves typing, this can *also* collide real on-disk paths versus aliases, i.e. what if there's a top-level directory named `test` vs `library/test`?). After #133492, `PathSet::check` matches on *prefixes* as well. There are of course some more complicated exceptions, special handling and `suite_path` logic differences, but this is generally an issue because...

Both `CodegenGCC` and `CodegenCranelift` test `Step`s have `should_run` conditions:

https://github.com/rust-lang/rust/blob/2061630860906083681a32ed748e53bee4a03fb6/src/bootstrap/src/core/build_steps/test.rs#L3406-L3408

https://github.com/rust-lang/rust/blob/2061630860906083681a32ed748e53bee4a03fb6/src/bootstrap/src/core/build_steps/test.rs#L3530-L3532

These test `Step`s are registered here via `describe!`:

https://github.com/rust-lang/rust/blob/2061630860906083681a32ed748e53bee4a03fb6/src/bootstrap/src/core/builder/mod.rs#L925-L926

When the cli receives the filter string `compiler`, seemingly both test `Step`s would be eligible. However, there is a problem. These two test `Step`'s `should_run`s don't use `suite_path`, they use `run.paths`. This means that they are not eligble for `suite_path` special handling that will try to exercise all the test suites:

https://github.com/rust-lang/rust/blob/2061630860906083681a32ed748e53bee4a03fb6/src/bootstrap/src/core/builder/mod.rs#L465-L475

Instead, the Codegen{Cranelift,Gcc} test `Step`s go through the non-suite-path handling, which has "remove-PathSet-once-matched" semantics, so the first one to match (first `Step` to be registered in the test `describe!` that is eligible to match the `PathSet` will "eat" the `PathSet`) prevents any other non-suite-path test `Step`s that are registered later from being matched.

https://github.com/rust-lang/rust/blob/2061630860906083681a32ed748e53bee4a03fb6/src/bootstrap/src/core/builder/mod.rs#L489-L490

Indeed, if you inject some printf logging, you can see how the `PathSet` is consumed:

```
[07:49] Joe:rust | ./x test compiler --dry-run
Building bootstrap
   Compiling bootstrap v0.0.0 (/home/joe/repos/rust/src/bootstrap)
    Finished `dev` profile [unoptimized] target(s) in 1.76s
StepDescription::run: paths=["compiler"]
StepDescription::run: handle all `PathSet`s: desc=StepDescription { default: true, only_hosts: true, should_run: 0x55e3fe0ece60, make_run: 0x55e3fe0ece80, name: "bootstrap::core::build_steps::test::CodegenCranelift", kind: Test }, pathsets=[Set({test::compiler/rustc_codegen_cranelift})]
StepDescription::maybe_run: pathsets=[Set({test::compiler/rustc_codegen_cranelift})]
CodegenCranelift
Building stage0 library artifacts (x86_64-unknown-linux-gnu)
Building compiler artifacts (stage0 -> stage1, x86_64-unknown-linux-gnu)
Creating a sysroot for stage1 compiler (use `rustup toolchain link 'name' build/host/stage1`)
cranelift not in rust.codegen-backends. skipping
StepDescription::run: handle all `PathSet`s: desc=StepDescription { default: false, only_hosts: false, should_run: 0x55e3fe4b8690, make_run: 0x55e3fe4b86b0, name: "bootstrap::core::build_steps::toolstate::ToolStateCheck", kind: Test }, pathsets=[]
StepDescription::run: handle all `PathSet`s: desc=StepDescription { default: true, only_hosts: true, should_run: 0x55e3fe0dad90, make_run: 0x55e3fe0daed0, name: "bootstrap::core::build_steps::test::Tidy", kind: Test }, pathsets=[]
[... some entries trimmed ...]
"bootstrap::core::build_steps::test::CodegenGCC", kind: Test }, pathsets=[]
```

Why does moving `test::CodegenGCC` and `test::CodegenCranelift` seemingly unblock `./x test compiler`? Well, it's because `test::CrateLibrustc` is now the non-suite-path test `Step` that first gets to eat the `PathSet`:

```
StepDescription::run: handle all `PathSet`s: desc=StepDescription { default: true, only_hosts: true, should_run: 0x5610308a9ef0, make_run: 0x5610308a9f40, name: "bootstrap::core::build_steps::test::CrateLibrustc", kind: Test }, pathsets=[Set({test::compiler})]
[... some entries trimmed ...]
StepDescription::run: handle all `PathSet`s: desc=StepDescription { default: true, only_hosts: true, should_run: 0x5610308afe60, make_run: 0x5610308afe80, name: "bootstrap::core::build_steps::test::CodegenCranelift", kind: Test }, pathsets=[]
StepDescription::run: handle all `PathSet`s: desc=StepDescription { default: true, only_hosts: true, should_run: 0x5610308b0d20, make_run: 0x5610308b0d40, name: "bootstrap::core::build_steps::test::CodegenGCC", kind: Test }, pathsets=[]
```

So now `CodegenCranelift` no longer tries to run because the `PathSet` got eaten by an earlier-registered test `Step`.

Ok. Now you may ask: how does https://github.com/rust-lang/rust/pull/133492 regress `./x test compiler`? Because before #133492, `PathSet`-based eligibility check only matches *suffixes*, but now #133492 also make `PathSet`-based eligibility also matches *prefixes*, making #133492 a *necessary* condition to break `./x test compiler` but is not the "root cause", in my honest opinion. The *sufficient* condition is arguably the `PathSet`-related handling itself.

I'm also only describing why this is a *problem*, I don't quite have a good solution in mind just yet.

(Also I'm not sure why there's a specific distinction of `suite_path`, I suspect that's only used for `compiletest`-managed `tests/`.)

**Review by jieyouxu — APPROVED** [review]
In any case, this is a sufficient fix for the basic workflow of testing the compiler unit crates. Thanks.

**Pull request merged**